src/Psalm/Internal/Type/TypeParser.php

```diff
@@ -1692,4 +1691,50 @@ private static function getTypeFromKeyedArrays(
             $from_docblock,
         );
     }
+
+
+    /**
+     * @param non-empty-array<Atomic> $types
+     * @return non-empty-array<Atomic>
+     */
+    private static function resolveTemplatedTypes(array $types): array
+    {
+        $resolved_types = [];
+        foreach ($types as $type) {
+            if ($type instanceof TTemplateParam) {
+                $resolved_types[] = self::resolveTemplatedTypes($type->as->getAtomicTypes());
+                continue;
+            }
+
+            $resolved_types[] = [$type];
+        }
+
+        $resolved_types = array_merge(...$resolved_types);
+        assert($resolved_types !== []);
+        return $resolved_types;
+    }
+
+    /**
+     * @param non-empty-array<Atomic> $types
```

Comment:
I've tried to add `@psalm-assert-if-true` so that we can provide `non-empty-array<TKeyedArray>` to get rid of the inline `@var` annotations above but that sadly didn't work but I dont have time to improve this for now.